### ISSUE TITLE
Panic 'byte index 1 is out of bounds of ``'

### ISSUE THREAD

**Issue by Hocuri** [dignifiedquire/rust-escaper#3]
Failing test:

```
    #[test]
    fn test_decode_panic() {
        let bytes = b"&#1
 ;";
        escaper::decode_html_buf_sloppy(bytes).ok();
    }
```

```
running 1 test
thread 'mimeparser::tests::test_decode_panic' panicked at 'byte index 1 is out of bounds of ``', /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/escaper-0.1.0/src/decode.rs:168:53
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/panicking.rs:92:14
   2: core::str::slice_error_fail
   3: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::RangeFrom<usize>>::index
   4: core::str::traits::<impl core::ops::index::Index<I> for str>::index
   5: <alloc::string::String as core::ops::index::Index<core::ops::range::RangeFrom<usize>>>::index
   6: escaper::decode::decode_html_rw
   7: escaper::decode::decode_html_buf_sloppy
```

**Comment by Jikstra** [comment]
Seems like already this explodes:
```
let mut buf = String::with_capacity(8);
println!("{:?}", &buf[1..]);
```

**Comment by Hocuri** [comment]
What you are doing in these two lines is, create an empty string (`''`) and directly try to access [1..].

The issue here is not that it's impossible to access index 1 of an empty string, but that it's tried in the first place. For most inputs, some char is pushed before, but in that case, this seems not to be the case.

Indexing is probably a bad idea anyway in Rust as, well, it can panic.

**Comment by Jikstra** [comment]
I'm currently trying to understand what's happening in the decoder part. The `&buf[1..]` creates a String slice where the first character is getting skipped. Which of course cannot happen if buf is empty. Maybe we can move the logic into the `decode_numeric` method? So first checking that the length is at least greater then 1, then slice it, then progress..

EDIT: Created a pr around this idea. Of course I don't know shit about all this. 

> Indexing is probably a bad idea anyway in Rust as, well, it can panic.

Is there a better way? I quick duck only showed rfcs and proposals ^^

**Comment by link2xt** [comment]
I made a fix #5.

The problem with parsing `&#1\n;` is that on newline after `1` the state was not reverted to `Normal`, but `buf` was cleared. And when reading `;` we are still in the `Numeric` state and try to read from `buf` which is empty after reading erroneous whitespace.

My solution is to revert to `Normal` state when we are in a "sloppy" mode and invalid character is encountered. `;` is read as normal character in this case.
